### PR TITLE
WIP: Create Signup Screen UI

### PR DESCRIPTION
**Pull request by muammerakca10**

### CODE REVIEW

**Review by safaorhan — CHANGES_REQUESTED** [review]
First iteration of my feedback is done!

Try to address all of my comments and try to find areas where you could apply the same suggestions.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        //Password Error Messages Start
+        fun passwordErrorMessages(password: String) = when {
+            !passwordNullOrEmptyControl(password) -> "Password is required."
+            !passwordDigitControl(password) -> "Password must contain one digit, one uppercase letter, one lowercase letter and one special character."
+            !passwordUpperCaseControl(password) -> "Password must contain one digit, one uppercase letter, one lowercase letter and one special character."
+            !passwordLowerCaseControl(password) -> "Password must contain one digit, one uppercase letter, one lowercase letter and one special character."
+            !passwordLetterOrDigitCharacterControl(password) -> "Password must contain one digit, one uppercase letter, one lowercase letter and one special character."
+            !passwordMinCharacterControl(password) -> "Password is too short."
+            !passwordMaxCharacterControl(password) -> "Password is too long."
+            !passwordSpaceControl(password) -> "Password cannot contain space."
```
Comment (cmfndkl) [inline app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:133]: User visible strings should always be extracted into strings.xml.

File: app/src/main/res/layout/activity_sign_up.xml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    <Button
+        android:id="@+id/alreadySignButton"
+        style="@style/Widget.MaterialComponents.Button.TextButton"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:layout_marginTop="32dp"
+        android:text="@string/already_account"
+        android:textColor="@color/light_green"
+        android:textSize="14sp" />
```
Comment (cmfndkl) [inline app/src/main/res/layout/activity_sign_up.xml:85]: All buttons must be MaterialButtons.

Comment (safaorhan) [reply app/src/main/res/layout/activity_sign_up.xml:82]: All buttons must be MaterialButtons that's correct. However, if the theme of the app extends from a MaterialComponents theme, then Button -> MaterialButton conversion is done automatically.

Check: https://stackoverflow.com/questions/57925853/is-any-difference-between-a-materialbutton-and-a-simple-button

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -9,7 +10,10 @@
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/Theme.CactusNotes">
-        <activity android:name=".MainActivity">
+        <activity
+            android:name="com.muammerakca.cactusnotes.SignUpActivity"
+            android:label="@string/signUp_activity_name">
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/AndroidManifest.xml:15]: Done

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        
+        val textInputEmail: TextInputLayout = findViewById(R.id.textInputEmail)
+        val textInputUsername: TextInputLayout = findViewById(R.id.textInputUsername)
+        val textInputPassword: TextInputLayout = findViewById(R.id.textInputPassword)
+        val signupButton: MaterialButton = findViewById(R.id.signupButton)
+        val alreadySignButton: Button = findViewById(R.id.alreadySignButton)
+
+        //-------------------------------------------------------------------------------------------
+
+        //Email Control Start
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:23]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            return email.length > 5
+        }
+
+        fun emailMaxCharacterControl(email: String): Boolean {
+            return email.length < 50
+        } // Email Control End
+
+        //Email Error Messages Start
+        fun emailErrorMessages(email: String) = when {
+            !emailNullEmptyControl(email) -> "Email is required."
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:39]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        //-------------------------------------------------------------------------------------------
+
+        //Username Controls Start
+        fun usernameNullEmptyControl(username: String): Boolean {
+            return !username.isEmpty()
+        }
+
+        fun usernameEligibleCharactersControl(username: String): Boolean {
+            return username.all { it.isDigit() || it.isLowerCase() || it == '_' }
+            //username!!.contains("[a-z][0-9]_") seklinde de yazilabilir mi?
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+            !passwordSpaceControl(password) -> "Password cannot contain space."
+            else -> null
+        } //Password Error Messages Finish
+
+        //-------------------------------------------------------------------------------------------
+
+        //Button Click Start
+        signupButton.setOnClickListener() {
+            var email =
+                textInputEmail.editText!!.text.toString() //bosluklara hassasiyeti artirmak icin trimleri kaldirdim
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/layout/activity_sign_up.xml (hunk 2 above)
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml:82]: Thank you for feedbacks.

File: app/src/main/res/layout/activity_sign_up.xml [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+        <com.google.android.material.textfield.TextInputEditText
+
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content" />
+
+    </com.google.android.material.textfield.TextInputLayout>
+
+    <com.google.android.material.textfield.TextInputLayout
+
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/layout/activity_sign_up.xml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:hint="@string/email_hint"
+        app:errorEnabled="false">
+
+        <com.google.android.material.textfield.TextInputEditText
+
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content" />
+
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml:24]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -1,3 +1,10 @@
 <resources>
-    <string name="app_name">CactusNotes</string>
+    <string name="app_name">Cactus Notes</string>
+    <string name="signupActivity">Sign up</string>
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/res/values/strings.xml]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/layout/activity_sign_up.xml [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        style="@style/Widget.MaterialComponents.Button.TextButton"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:layout_marginTop="32dp"
+        android:text="@string/already_account"
+        android:textColor="@color/light_green"
+        android:textSize="14sp" />
+
+
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/layout/activity_sign_up.xml [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:hint="@string/password_hint"
+        app:errorEnabled="false"
+        app:passwordToggleEnabled="true">
+
+        <com.google.android.material.textfield.TextInputEditText
+
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt (hunk 1 above)
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+            !usernameMinCharacterControl(username) -> "Username is too short."
+            !usernameMaxCharacterControl(username) -> "Username is too long."
+            else -> null
+
+        } //Username Error Messages Finish
+
+        //-------------------------------------------------------------------------------------------
+
+        //Password Controls Start
+        fun passwordNullOrEmptyControl(password: String): Boolean {
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_sign_up)
+
+        val textInputEmail: TextInputLayout = findViewById(R.id.textInputEmail)
+        val textInputUsername: TextInputLayout = findViewById(R.id.textInputUsername)
+        val textInputPassword: TextInputLayout = findViewById(R.id.textInputPassword)
+        val signupButton: MaterialButton = findViewById(R.id.signupButton)
+        val alreadySignButton: Button = findViewById(R.id.alreadySignButton)
+
+        //Email Error Messages Start
```
Comment (safaorhan) [inline app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:21]: We can remove these comment lines

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        fun emailErrorMessages(email: String) = when {
+            email.isEmpty() -> getString(R.string.email_required)
+            !email.contains(".") || !email.contains("@") || email.length <= 5 || email.length >= 50 -> getString(R.string.email_invalid)
+            email.contains(" ") -> getString(R.string.email_cannot_space)
+            else -> null
+        } //Email Error Messages Finish
```
Comment (safaorhan) [inline app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:27]: We can remove these comment lines

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+        fun passwordErrorMessages(password: String) = when {
+            password.isEmpty() -> getString(R.string.password_required)
+            !password.any { it.isDigit() } || !password.any { it.isDigit() } || !password.any { it.isLowerCase() } || !password.any { it.isLetterOrDigit() } -> getString(R.string.password_must_contains_character)
+            password.length < 8 -> getString(R.string.password_short)
+            password.length >= 40 -> getString(R.string.password_long)
+            password.contains(" ") -> getString(R.string.password_cannot_space)
+            else -> null
+        } //Password Error Messages Finish
+
+        //Button Click Start
```
Comment (safaorhan) [inline app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:48]: We can remove these comment lines

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+        setContentView(R.layout.activity_sign_up)
+
+        val textInputEmail: TextInputLayout = findViewById(R.id.textInputEmail)
+        val textInputUsername: TextInputLayout = findViewById(R.id.textInputUsername)
+        val textInputPassword: TextInputLayout = findViewById(R.id.textInputPassword)
+        val signupButton: MaterialButton = findViewById(R.id.signupButton)
+        val alreadySignButton: Button = findViewById(R.id.alreadySignButton)
+
+        //Email Error Messages Start
+        fun emailErrorMessages(email: String) = when {
```
Comment (safaorhan) [inline app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:22]: We can rename this into `validateEmail(email: String)`

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+            email.isEmpty() -> getString(R.string.email_required)
+            !email.contains(".") || !email.contains("@") || email.length <= 5 || email.length >= 50 -> getString(R.string.email_invalid)
+            email.contains(" ") -> getString(R.string.email_cannot_space)
+            else -> null
+        } //Email Error Messages Finish
+
+        //Username Error Messages Start
+        fun usernameErrorMessages(username: String) = when {
```
Comment (safaorhan) [inline app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:30]: This as well, validateUsername...

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+            username.isEmpty() -> getString(R.string.username_required)
+            !username.all { it.isDigit() || it.isLowerCase() || it == '_' } -> getString(R.string.username_only_this_characters)
+            username.length <= 2 -> getString(R.string.username_short)
+            username.length >= 20 -> getString(R.string.username_long)
+            else -> null
+        } //Username Error Messages Finish
+
+        //Password Error Messages Start
+        fun passwordErrorMessages(password: String) = when {
```
Comment (safaorhan) [inline app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:39]: ...validatePassword...

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+            password.isEmpty() -> getString(R.string.password_required)
+            !password.any { it.isDigit() } || !password.any { it.isDigit() } || !password.any { it.isLowerCase() } || !password.any { it.isLetterOrDigit() } -> getString(R.string.password_must_contains_character)
```
Comment (safaorhan) [inline app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:41]: 1. `!password.any { it.isDigit() }` is checked twice.
2. We can extract those into functions for better readability

Comment (safaorhan) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:41]: `fun String.containsDigit() = any { it.isDigit() }`
...
`!password.containsDigit() || !p...`

Comment (safaorhan) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:41]: `!password.any { it.isLetterOrDigit() }` checks if it does not contain any "letter or digit" but instead you should check:

`!password.any { !it.isLetterOrDigit() }` checks if it does not contain any character which is not "letter or digit"

File: app/src/main/res/layout/activity_sign_up.xml [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+        android:layout_marginStart="16dp"
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:hint="@string/username_hint">
+
+        <com.google.android.material.textfield.TextInputEditText
+
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content" />
+
```
Comment (safaorhan) [inline app/src/main/res/layout/activity_sign_up.xml:41]: No new-lines before closing tag

File: app/src/main/res/layout/activity_sign_up.xml [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:hint="@string/password_hint"
+        app:errorEnabled="false"
+        app:passwordToggleEnabled="true">
+
+        <com.google.android.material.textfield.TextInputEditText
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content" />
+
```
Comment (safaorhan) [inline app/src/main/res/layout/activity_sign_up.xml:59]: No new-lines before closing tag

File: app/src/main/res/values/strings.xml [hunk 23]
```diff
@@ -1,3 +1,23 @@
 <resources>
-    <string name="app_name">CactusNotes</string>
+    <string name="app_name">Cactus_Notes</string>
+    <string name="signupActivity">Sign_Up</string>
```
Comment (safaorhan) [inline app/src/main/res/values/strings.xml:3]: casing and fix user visible strings

**Review by safaorhan — CHANGES_REQUESTED** [review]
Well done! 🥳 

There are some critical and some minor problems, we should do another iteration!

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt (hunk 14 above)
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:21]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt (hunk 15 above)
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:27]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt (hunk 17 above)
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:22]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt (hunk 18 above)
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:30]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt (hunk 19 above)
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:39]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/layout/activity_sign_up.xml (hunk 21 above)
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml:41]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/layout/activity_sign_up.xml (hunk 22 above)
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml:59]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/values/strings.xml (hunk 23 above)
Comment (muammerakca10) [reply app/src/main/res/values/strings.xml:3]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt (hunk 16 above)
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:48]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/layout/activity_sign_up.xml [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:hint="@string/email_hint"
+        app:errorEnabled="false">
+
+        <com.google.android.material.textfield.TextInputEditText
+
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/res/layout/activity_sign_up.xml [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+        style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="16dp"
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:hint="@string/username_hint">
+
+        <com.google.android.material.textfield.TextInputEditText
+
```
[the remark this answers is not in the document]
Comment (muammerakca10) [reply app/src/main/res/layout/activity_sign_up.xml:38]: Thank you for feedback. I made the necessary corrections.

File: app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt (hunk 20 above)
Comment (muammerakca10) [reply app/src/main/java/com/muammerakca/cactusnotes/SignUpActivity.kt:41]: Thank you for feedback. I made the necessary corrections.

**Review by safaorhan — APPROVED** [review]
Looks good to me! We can merge this after this gets QAed

**Comment by cmfndkl** [comment]
Congratulations, nice job. However, you have to make some changes: 
- Password visibility is default enabled. You should make your password visibility invisible by default.  
- The "Already Have Account" button should be changed to "material button" and should be the same width as the "SIGN UP" button.

**Comment by safaorhan** [comment]
@cmfndkl we already discussed the material button thing earlier.

If you use a MaterialComponents theme, Button -> MaterialButton conversion is done automatically. So no need to use MaterialButtons in the XML explicitly.

**Review by safaorhan — APPROVED** [review]

**Pull request merged**